jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,828 @@
+    return;
+  if(expr.type() == java_boolean_type())
+  {
+    if(json.is_true())
+      block.add(code_assignt{expr, true_exprt{}});
+    else
+      block.add(code_assignt{expr, false_exprt{}});
+  }
+  else if(
+    expr.type() == java_int_type() || expr.type() == java_byte_type() ||
+    expr.type() == java_short_type() || expr.type() == java_long_type())
+  {
+    block.add(
+      code_assignt{expr, from_integer(std::stoi(json.value), expr.type())});
+  }
+  else if(expr.type() == java_double_type())
+  {
+    ieee_floatt ieee_float(to_floatbv_type(expr.type()));
+    ieee_float.from_double(std::stod(json.value));
+    block.add(code_assignt{expr, ieee_float.to_expr()});
+  }
+  else if(expr.type() == java_float_type())
+  {
+    ieee_floatt ieee_float(to_floatbv_type(expr.type()));
+    ieee_float.from_float(std::stof(json.value));
+    block.add(code_assignt{expr, ieee_float.to_expr()});
+  }
+  else if(expr.type() == java_char_type())
+  {
+    const std::wstring wide_value = utf8_to_utf16_native_endian(json.value);
+    PRECONDITION(wide_value.length() == 1);
+    block.add(
+      code_assignt{expr, from_integer(wide_value.front(), expr.type())});
+  }
+}
+
+/// One of the base cases of the recursive algorithm. See
+/// \ref assign_from_json_rec.
+static void assign_null(const exprt &expr, code_blockt &block)
+{
+  block.add(
+    code_assignt{expr, null_pointer_exprt{to_pointer_type(expr.type())}});
+}
+
+/// In the case of an assignment of an array given a JSON representation, this
+/// function assigns the data component of the array, which contains the array
+/// elements. \p expr is a pointer to the array containing the component.
+static void assign_array_data_component_from_json(
+  const exprt &expr,
+  const jsont &json,
+  const optionalt<std::string> &type_from_array,
+  det_creation_infot &info)
+{
+  const auto &java_class_type = followed_class_type(expr, info.symbol_table);
+  const auto &components = java_class_type.components();
+  const auto &element_type = static_cast<const typet &>(
+    to_pointer_type(expr.type()).subtype().find(ID_element_type));
+  const exprt data_member_expr = typecast_exprt::conditional_cast(
+    member_exprt{dereference_exprt{expr}, "data", components[2].type()},
+    pointer_type(element_type));
+
+  const symbol_exprt &array_init_data =
+    info.allocate_objects.allocate_automatic_local_object(
+      data_member_expr.type(), "det_array_data_init");
+  const code_assignt data_assign{array_init_data, data_member_expr, info.loc};
+  info.block.add(data_assign);
+
+  size_t index = 0;
+  const optionalt<std::string> inferred_element_type =
+    element_type_from_array_type(json, type_from_array);
+  const json_arrayt json_array = get_untyped_array(json, element_type);
+  for(auto it = json_array.begin(); it < json_array.end(); it++, index++)
+  {
+    const dereference_exprt element_at_index = array_element_from_pointer(
+      array_init_data, from_integer(index, java_int_type()));
```

Comment:
Took me a while to figure this one out myself. It's because we're dealing with a pointer array here, i.e. the array is represented as a pointer to its first element, and other elements can be accessed with pointer offsets (the `plus_exprt` in the definition of `array_element_from_pointer`). `index_exprt`s represent "actual" arrays. Java arrays are always represented as pointer arrays.